File: src/test/java/no/nav/fo/veilarbregistrering/sykemelding/SykemeldingServiceTest.java
package no.nav.fo.veilarbregistrering.sykemelding;

import org.junit.Test;

import java.time.LocalDate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class SykemeldingServiceTest {

    private SykemeldingService sykemeldingService;

    @Test
    public void hentSykmeldtInfoData_skal_håndtere_maksdato_lik_null() {
        SykemeldingGateway sykemeldingGateway = mock(SykemeldingGateway.class);
        when(sykemeldingGateway.hentReberegnetMaksdato(any())).thenReturn(Maksdato.nullable());

        sykemeldingService = new SykemeldingService(sykemeldingGateway);

        SykmeldtInfoData sykmeldtInfoData = sykemeldingService.hentSykmeldtInfoData(any());
        assertThat(sykmeldtInfoData.getMaksDato()).isNull();
        assertThat(sykmeldtInfoData.isErArbeidsrettetOppfolgingSykmeldtInngangAktiv()).isFalse();
    }

    @Test
    public void hentSykmeldtInfoData_skal_håndtere_maksdato() {
        SykemeldingGateway sykemeldingGateway = mock(SykemeldingGateway.class);
        when(sykemeldingGateway.hentReberegnetMaksdato(any())).thenReturn(Maksdato.of("2020-11-01"));

        sykemeldingService = new SykemeldingService(sykemeldingGateway);

        SykmeldtInfoData sykmeldtInfoData = sykemeldingService.hentSykmeldtInfoData(any());
        assertThat(sykmeldtInfoData.getMaksDato()).isEqualTo("2020-11-01");
        assertThat(sykmeldtInfoData.isErArbeidsrettetOppfolgingSykmeldtInngangAktiv()).isFalse();
    }

    @Test
    public void hentSykmeldtInfoData_skal_håndtere_maksdato_mellom_39_og_52_uker() {
        SykemeldingGateway sykemeldingGateway = mock(SykemeldingGateway.class);
        String maksdato = LocalDate.now().plusWeeks(3).toString();
        when(sykemeldingGateway.hentReberegnetMaksdato(any())).thenReturn(Maksdato.of(maksdato));

        sykemeldingService = new SykemeldingService(sykemeldingGateway);

        SykmeldtInfoData sykmeldtInfoData = sykemeldingService.hentSykmeldtInfoData(any());
        assertThat(sykmeldtInfoData.getMaksDato()).isEqualTo(maksdato);
        assertThat(sykmeldtInfoData.isErArbeidsrettetOppfolgingSykmeldtInngangAktiv()).isTrue();
    }
}

File: src/test/java/no/nav/fo/veilarbregistrering/sykemelding/SykemeldingServiceTest.kt
package no.nav.fo.veilarbregistrering.sykemelding

import io.mockk.every
import io.mockk.mockk
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDate

class SykemeldingServiceTest {
    private lateinit var sykemeldingService: SykemeldingService
    private lateinit var sykemeldingGateway: SykemeldingGateway

    @BeforeEach
    fun setup() {
        sykemeldingGateway = mockk()
        sykemeldingService = SykemeldingService(
            sykemeldingGateway,
            mockk(relaxed = true)
        )
    }

    @Test
    fun `hentSykmeldtInfoData skal håndtere maksdato lik null`() {
        every { sykemeldingGateway.hentReberegnetMaksdato(any()) } returns Maksdato.nullable()
        val sykmeldtInfoData = sykemeldingService.hentSykmeldtInfoData(FNR)
        Assertions.assertThat(sykmeldtInfoData.getMaksDato()).isNull()
        Assertions.assertThat(sykmeldtInfoData.isErArbeidsrettetOppfolgingSykmeldtInngangAktiv).isFalse
    }

    @Test
    fun `hentSykmeldtInfoData skal håndtere maksdato`() {
        val datoUtenforAktivInngang = LocalDate.now().plusWeeks(20)
        every { sykemeldingGateway.hentReberegnetMaksdato(any()) } returns Maksdato.of(datoUtenforAktivInngang.toString())
        val sykmeldtInfoData = sykemeldingService.hentSykmeldtInfoData(FNR)
        Assertions.assertThat(sykmeldtInfoData.getMaksDato()).isEqualTo(datoUtenforAktivInngang.toString())
        Assertions.assertThat(sykmeldtInfoData.isErArbeidsrettetOppfolgingSykmeldtInngangAktiv).isFalse
    }

    @Test
    fun `hentSykmeldtInfoData skal håndtere maksdato mellom 39 og 52 uker`() {
        val maksdato = LocalDate.now().plusWeeks(3).toString()
        every { sykemeldingGateway.hentReberegnetMaksdato(any()) } returns Maksdato.of(maksdato)

        val sykmeldtInfoData = sykemeldingService.hentSykmeldtInfoData(FNR)
        Assertions.assertThat(sykmeldtInfoData.getMaksDato()).isEqualTo(maksdato)
        Assertions.assertThat(sykmeldtInfoData.isErArbeidsrettetOppfolgingSykmeldtInngangAktiv).isTrue
    }

    companion object {
        private val FNR = Foedselsnummer.of("121212123123")
    }
}